hil/ext/switches/dellnos9.py

```diff
@@ -0,0 +1,380 @@
+
+        if channel == 'vlan/native':
+            if new_network is None:
+                self._remove_native_vlan(interface)
+            else:
+                self._set_native_vlan(interface, new_network)
+        else:
+            vlan_id = channel.replace('vlan/', '')
+            legal = get_network_allocator(). \
+                is_legal_channel_for(channel, vlan_id)
+            assert legal, "HIL passed an invalid channel to the switch!"
+
+            if new_network is None:
+                self._remove_vlan_from_trunk(interface, vlan_id)
+            else:
+                assert new_network == vlan_id
+                self._add_vlan_to_trunk(interface, vlan_id)
+
+    def revert_port(self, port):
+        self._remove_all_vlans_from_trunk(port)
+        if self._get_native_vlan(port) is not None:
+            self._remove_native_vlan(port)
+
+    def get_port_networks(self, ports):
+        response = {}
+        for port in ports:
+            response[port] = self._get_vlans(port.label)
+            native = self._get_native_vlan(port.label)
+            if native is not None:
+                response[port].append(native)
+
+        return response
+
+    def _get_vlans(self, interface):
+        """ Return the vlans of a trunk port.
+
+        Does not include the native vlan. Use _get_native_vlan.
+
+        Args:
+            interface: interface to return the vlans of
+
+        Returns: List containing the vlans of the form:
+        [('vlan/vlan1', vlan1), ('vlan/vlan2', vlan2)]
+        """
+
+        # It uses the REST API CLI which is slow but it is the only way
+        # because the switch is VLAN centric. Doing a GET on interface won't
+        # return the VLANs on it, we would have to do get on all vlans (if that
+        # worked reliably in the first place) and then find our interface there
+        # which is not feasible.
+
+        response = self._get_port_info(interface).splitlines()
+
+        try:
+            index = response.index("Vlanmembership:") + 3
+        except ValueError:
+            # happens when our response has almost no info on a switchport
+            # because it's off. the response lacks the word "vlanmembership"
+            return []
+        if response[index] == '':
+            # a port with no tagged interface has no section with "T"
+            return []
+        vlan_list = response[index].replace('T', '').split(',')
+        return [('vlan/%s' % x, x) for x in vlan_list]
+
+    def _get_native_vlan(self, interface):
+        """ Return the native vlan of an interface.
+
+        Args:
+            interface: interface to return the native vlan of
+
+        Returns: Tuple of the form ('vlan/native', vlan) or None
+
+        Similar to _get_vlans()
+        """
+        response = self._get_port_info(interface)
+        if response.find('NativeVlanId:') == -1:
+            return None
+        begin = response.find('NativeVlanId:') + len('NativeVlanId:')
```

Comment:
Use a regular expression to figure out if the native vlan exists and to get its id.

r'NativeVlanId:(\d+)\.'